[AArch64] Lower extending sitofp using tbl

momchil-velikov opened the pull request:
In a similar manner as in https://reviews.llvm.org/D133494
use `TBL` to place bytes in the *upper* part of `i32` elements
and then convert to float using fixed-point `scvtf`, i.e.

    scvtf Vd.4s, Vn.4s, #24

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -17840,13 +17887,39 @@ static SDValue performVectorCompareAndMaskUnaryOpCombine(SDNode *N,
+    return SDValue();
+
+  SDNode *VASHR = N->getOperand(0).getNode();
+  if (VASHR->getOpcode() != AArch64ISD::VASHR ||
```
Im not sure if this should be matched from shifts in general. Can we reinstate https://github.com/llvm/llvm-project/pull/91924/files but using fmul and match using that? It would be nice if it could use tablegen like https://reviews.llvm.org/D113199 too, if that is possible.

davemgreen replied:
> Im not sure if this should be matched from shifts in general

I think the shift would need to be exact. https://alive2.llvm.org/ce/z/F9usML. From what I can tell we are trying to add this: https://alive2.llvm.org/ce/z/82h6nn, where the shift would be exact, but that wouldn't be true for other cases.

It would probably be good to get the scvtf_fixed working in one patch first with some testing, and have the tbl loop transform in a second.

momchil-velikov replied:
Ack.

momchil-velikov replied:
ISel patterns forked out to https://github.com/llvm/llvm-project/pull/92922 and now require the shift to be exact.

momchil-velikov committed:
[AArch64] Lower extending sitofp using tbl

In a similar manner as in https://reviews.llvm.org/D133494
use `TBL` to place bytes in the *upper* part of `i32` elements
and then convert to float using fixed-point `scvtf`, i.e.

    scvtf Vd.4s, Vn.4s, #24

Change-Id: Ib9df3e4243612cbee8560907b24b14e76b61f265

momchil-velikov commented:
Ping?

davemgreen reviewed (approved):
Considering this operates on sitofp I think it should be OK. LGTM

momchil-velikov committed:
fixup: add a missing CHECK line

momchil-velikov commented on llvm/test/CodeGen/AArch64/sitofp-to-tbl.ll:
```diff
@@ -0,0 +1,196 @@
+; CHECK-NEXT:  .byte    255
+; CHECK-NEXT:  .byte    255
+; CHECK-NEXT:  .byte    255
+
```
Fixed.

momchil-velikov committed:
[fixup] Add some more tests

momchil-velikov commented on llvm/test/CodeGen/AArch64/sitofp-to-tbl.ll:
```diff
@@ -0,0 +1,197 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -verify-machineinstrs < %s | FileCheck %s
+
+target triple = "aarch64-linux"
```
I added some tests.

scweng commented:
Hi, at Google we are currently seeing two test failures after this commit:

1. In Jax, [lax_numpy_test](https://github.com/google/jax/blob/main/tests/lax_numpy_test.py) has some floating point number differences like -0.75 vs -0.6875 and 0.5 vs 0.6875.
2. In an internal test case for Eigen3, -4.22987 vs -5.

We are still trying to reduce the Eigen3 test, which is pure C++. I know that this change was landed over a week ago, but with the two differences both point to this commit, is it possible to roll it back until further investigation?

scweng commented:
Thank you! I'm working on reducing the case.

scweng commented:
The .ii file is still over 500k bytes long because eigen is a header library. But I've seen the assembly diff boils down to two snippets of exactly this. Before this commit

```
          fmov    s0, w21
          mov     v0.s[1], w22
          shl     v0.2s, v0.2s, #24
          sshr    v0.2s, v0.2s, #24
          scvtf   v0.2s, v0.2s
          fcvtzs  w8, s0
          str     q0, [sp]                        // 16-byte Folded Spill
          cmp     w8, w21, sxtb
          b.ne    .LBB16_10
  // %bb.2:                               // %if.end
                                          //   in Loop: Header=BB16_1 Depth=1
          // ... function call removed ...
          ldr     q0, [sp]                        // 16-byte Folded Reload
          mov     s0, v0.s[1]
          fcvtzs  w8, s0
          cmp     w8, w22, sxtb
          b.ne    .LBB16_9
```

At this commit:

```
          mov     v0.b[3], w21
          mov     v0.b[7], w22
          scvtf   v0.2s, v0.2s, #24
          fcvtzs  w8, s0
          str     q0, [sp]                        // 16-byte Folded Spill
          cmp     w8, w21, sxtb
          b.ne    .LBB16_10
  // %bb.2:                               // %if.end
                                          //   in Loop: Header=BB16_1 Depth=1
          // ... function call removed ...
          ldr     q0, [sp]                        // 16-byte Folded Reload
          mov     s0, v0.s[1]
          fcvtzs  w8, s0
          cmp     w8, w22, sxtb
          b.ne    .LBB16_9
```

The other part is identical except for the input registers (`w20` and `w19` except `w21` and `w22`). Unified diff below:

```
--- array_cwise_no_sitofp_tbl.s 2024-06-29 02:48:01.129830833 +0800
+++ array_cwise_sitofp_tbl.s    2024-06-29 02:50:15.145362022 +0800
@@ -1296,11 +1296,9 @@
        bl      rand
        mov     w19, w0
        bl      rand
-       fmov    s0, w21
-       mov     v0.s[1], w22
-       shl     v0.2s, v0.2s, #24
-       sshr    v0.2s, v0.2s, #24
-       scvtf   v0.2s, v0.2s
+       mov     v0.b[3], w21
+       mov     v0.b[7], w22
+       scvtf   v0.2s, v0.2s, #24
        fcvtzs  w8, s0
        str     q0, [sp]                        // 16-byte Folded Spill
        cmp     w8, w21, sxtb
@@ -1317,13 +1315,11 @@
        b.ne    .LBB16_9
 // %bb.3:                               // %if.end.1
                                         //   in Loop: Header=BB16_1 Depth=1
-       fmov    s0, w20
+       mov     v0.b[3], w20
        ldr     x8, [x24, :lo12:.L_MergedGlobals+8]
        sub     x0, x8, #32
-       mov     v0.s[1], w19
-       shl     v0.2s, v0.2s, #24
-       sshr    v0.2s, v0.2s, #24
-       scvtf   v0.2s, v0.2s
+       mov     v0.b[7], w19
+       scvtf   v0.2s, v0.2s, #24
        str     q0, [sp]                        // 16-byte Folded Spill
        bl      _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE5c_strEv
        ldr     q0, [sp]                        // 16-byte Folded Reload
```

I'm not sure how the two snippets behave differently. And if they do behave differently, it feels like a latent bug uncovered by this commit?

scweng commented:
Oh, "MOV (from general)" used in the new assembly sequences does not clear other bits, while the shift left then right method does, doesn't it? That would explain that the differences seem to be precision loss.

momchil-velikov commented:
Thank you very much, that's incredibly helpful !

momchil-velikov commented:
Should be OK to re-commit after https://github.com/llvm/llvm-project/pull/97448
I ran Eigen tests, they look good (well, I get random distinct 0-3 failures on each run, which happens on `main` too).
Will re-commit next week, unless somebody spots a problem.
